### PR TITLE
[AIRFLOW-4939] Add default_task_retries config

### PR DESCRIPTION
**Pull request by msumit**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issue
  - https://issues.apache.org/jira/browse/AIRFLOW-4939

### Description

- [x] Adding a config level support to specify default task retries, which can be overridden at dag or task level. 

### Tests

- [x] My PR adds relevant unit test cases.

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: tests/operators/test_bash_operator.py [hunk 1]
```diff
@@ -98,3 +105,22 @@ def test_return_value(self):
...
+        self.assertEqual(bash_operator.retries, 2)
+
+    def test_default_retries(self):
+        bash_operator = BashOperator(
+            bash_command='echo "stdout"',
+            task_id='test_default_retries',
+            dag=None
+        )
+
+        self.assertEqual(bash_operator.retries, 3)
```
Comment (ashb) [inline tests/operators/test_bash_operator.py:121]: Does this test pass? Does default values are evaluated at module parse time, so changing the config value at runtime here won't affect the already-loaded default argument in the fn.

File: airflow/config_templates/default_airflow.cfg [hunk 2]
```diff
@@ -196,6 +196,9 @@ worker_precheck = False
 # When discovering DAGs, ignore any files that don't contain the strings `DAG` and `airflow`.
 dag_discovery_safe_mode = True
 
+# The no of retries each task is going to have by default. Can be overridden at dag or task level.
```
Comment (aoen) [inline airflow/config_templates/default_airflow.cfg]: s/no/number

**Comment by aoen** [comment]
LGTM after nit, tests pass, and default retries changed. For context to others: this is a pretty important change WRT reliability, since Airflow doesn't differentiate between Airflow exceptions/system failures and actual operator exceptions, which makes tasks with 0 retries quite unreliable.

File: airflow/config_templates/default_airflow.cfg (hunk 2 above)
Comment (aoen) [reply airflow/config_templates/default_airflow.cfg]: Oh and also default retries should be changed to 0.

File: tests/operators/test_bash_operator.py (hunk 1 above)
Comment (msumit) [reply tests/operators/test_bash_operator.py:121]: yup, you are correct, it doesn't pass. any pointers to fix this?

Comment (ashb) [reply tests/operators/test_bash_operator.py:121]: Either mess around with reloading modules (I think we do it in the logging tests?) or just don't test this?

File: tests/operators/test_bash_operator.py [hunk 3]
```diff
@@ -33,6 +34,15 @@
 
 
 class BashOperatorTest(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        configuration.conf.set('core', 'default_task_retries', "3")
```
Comment (milton0825) [inline tests/operators/test_bash_operator.py]: I think you can use https://github.com/apache/airflow/blob/577f8936727454eb4ff54588cd525c5244ef6d52/tests/test_utils/decorators.py#L28 instead

**Comment by msumit** [comment]
@ashb @milton0825 after mocking the tests are passing on my local, but still failing on Travis. Any further suggestions?

**Comment by ashb** [comment]
```
======================================================================
56) FAIL: test_default_retries (tests.operators.test_bash_operator.BashOperatorTest)
----------------------------------------------------------------------
   Traceback (most recent call last):
    /usr/lib/python3.5/unittest/mock.py line 1157 in patched
      return func(*args, **keywargs)
    tests/operators/test_bash_operator.py line 124 in test_default_retries
      self.assertEqual(bash_operator.retries, 3)
   AssertionError: 0 != 3
```

You still have the problem that python evaulates the default value ones at the time the module is first loaded so you will need to do something like [importlib.reload](https://docs.python.org/3/library/importlib.html#importlib.reload)

```python
    @mock.patch.object(configuration.conf, 'getint')
    def test_task_retries(self, mock_config):
        mock_config.return_value = '3'
        importlib.reload(airflow.operators.bash_operator)
        bash_operator = BashOperator(
            bash_command='echo "stdout"',
            task_id='test_task_retries',
            retries=2,
            dag=None
        )

        self.assertEqual(bash_operator.retries, 2)
```

And similar for the other tests you add. We should also add:

```python
    @classmethod
    def tearDownClass(cls):
            importlib.reload(airflow.operators.bash_operator)
```

to reset the default back to the default for any other tests.

Not that you will need to import the "whole" module to get the _module object_ to pass to `reload()`

**Comment by msumit** [comment]
@ashb tried `importlib.reload` as well, but unfortunately, it's not working as well. Should we just skip the failing test?

**Comment by msumit** [comment]
Ok, the bug I was hitting in tests was this (from [here](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments))
`Python’s default arguments are evaluated once when the function is defined, not each time the function is called (like it is in say, Ruby). This means that if you use a mutable default argument and mutate it, you will and have mutated that object for all future calls to the function as well.`

So I've changed the code a bit and now it's much cleaner than before and no hacks in tests as well. Please review it again and merge :).

**Comment by aoen** [comment]
LGTM

**Pull request merged**